aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -5,33 +5,152 @@
+  output_sizes[0] = input.size(0);
+  std::vector<int64_t> checked_signal_sizes(signal_ndim);
+  int64_t input_size;
+  for (int64_t i = 0; i < signal_ndim; i++) {
+    input_size = input.size(i + 1);
+    if (i == signal_ndim - 1 && onesided && complex_input && !complex_output) {
+      // if last dim and complex-to-real onesided, input is only half of
+      // signal, and we need to infer basing on signal_sizes, if given
+      int64_t inferred_size;
+      if (signal_sizes.size() > 0) {
+        inferred_size = infer_ft_complex_to_real_onesided_size(input_size, signal_sizes[i]);
+      } else {
+        inferred_size = infer_ft_complex_to_real_onesided_size(input_size);
+      }
+      checked_signal_sizes[i] = inferred_size;
+      output_sizes[i + 1] = inferred_size;
+    } else {
+      if (i == signal_ndim - 1 && onesided && !complex_input && complex_output) {
+        // if last dim and real-to-complex onesided, output should be only
+        // half of the signal, and we need to infer using input_size
+        output_sizes[i + 1] = infer_ft_real_to_complex_onesided_size(input_size);
+      } else {
+        output_sizes[i + 1] = input_size;
+      }
+      checked_signal_sizes[i] = input_size;
+      if (signal_sizes.size() > 0 && signal_sizes[i] != checked_signal_sizes[i]) {
+        std::ostringstream ss;
+        ss << "Expected given signal_sizes=" << signal_sizes << " to have same "
+           << "shape with input at signal dimension " << i << ", but got "
+           << "signal_sizes[" << i << "] = " << signal_sizes[i] << " and "
+           << "input.size(" << i + (int)is_batched << ") = " << input_size;
+        throw std::runtime_error(ss.str());
+      }
+    }
+  }
+  if (complex_output) {
+    output_sizes[signal_ndim + 1] = 2;
+  }
+
+  Tensor output = at::_fft_with_size(input, signal_ndim, complex_input,
+                                     complex_output, inverse,
+                                     checked_signal_sizes, normalized, onesided,
+                                     output_sizes);
+
+  // un-batch if needed
+  if (!is_batched) {
+    return output.squeeze_(0);
+  } else {
+    return output;
+  }
 }
 
+Tensor fft(const Tensor& self, const int64_t signal_ndim, const bool normalized) {
+  return _fft(self, signal_ndim, true, true, false, {}, normalized, false);
+}
+
+Tensor ifft(const Tensor& self, const int64_t signal_ndim, const bool normalized) {
+  return _fft(self, signal_ndim, true, true, true, {}, normalized, false);
+}
+
+Tensor rfft(const Tensor& self, const int64_t signal_ndim, const bool normalized,
+           const bool onesided) {
+  return _fft(self, signal_ndim, false, true, false, {}, normalized, onesided);
+}
+
+Tensor irfft(const Tensor& self, const int64_t signal_ndim,  IntList signal_sizes,
+            const bool normalized, const bool onesided) {
+  return _fft(self, signal_ndim, true, false, true, signal_sizes, normalized, onesided);
+}
```

Comment:
Not all boolean combinations are exercised here. Are they all valid or is the function written such that the given settings are the only valid ones?

Reply:
The given settings are the only valid ones here.

Reply:
Can we please use the `_fft(..., /*arg_name=*/true, /*arg_name2=*/false)` notation? That would make it much more readable. I'm concerned about having 3 booleans in a row